future-apps/payroll/contracts/Payroll.sol

```diff
@@ -112,119 +147,158 @@ contract Payroll is AragonApp { //, IForwarder { // makes coverage crash (remove
      */
     function setEmployeeSalary(
         uint128 employeeId,
-        uint256 yearlyDenominationSalary
+        uint256 denominationSalary
     )
+        employeeExists(employeeId)
         external
-        authP(ADD_EMPLOYEE_ROLE, arr(yearlyDenominationSalary, 0))
+        authP(SET_EMPLOYEE_SALARY_ROLE, arr(uint256(employeeId), denominationSalary))
     {
-        /* check that employee exists */
-        require(employeeIds[employees[employeeId].accountAddress] != 0);
+        uint64 timestamp = getTimestamp64();
 
-        employees[employeeId].denominationTokenSalary = yearlyDenominationSalary.toSecondDenominationToken();
+        // Add owed salary to employee's accrued value
+        uint256 owed = _getOwedSalary(employeeId, timestamp);
+
+        employees[employeeId].lastPayroll = timestamp;
+        _addAccruedValue(employeeId, owed);
+
+        // set new salary
+        employees[employeeId].denominationTokenSalary = denominationSalary;
+
+        emit SetEmployeeSalary(employeeId, denominationSalary);
+    }
+
+    /**
+     * @dev Terminate employee from Payroll
+     * @notice Terminate employee #`employeeId` from Payroll
+     * @param employeeId Employee's identifier
+     */
+    function terminateEmployeeNow(
+        uint128 employeeId
+    )
+        employeeExists(employeeId)
+        external
+        authP(TERMINATE_EMPLOYEE_ROLE, arr(uint256(employeeId)))
+    {
+        _terminateEmployee(employeeId, getTimestamp64());
     }
 
     /**
-     * @dev Remove employee from Payroll
-     * @notice Updates also global Payroll salary sum
+     * @dev Terminate employee from Payroll
+     * @notice Terminate employee #`employeeId` from Payroll
      * @param employeeId Employee's identifier
+     * @param endDate Termination date
      */
-    function removeEmployee(uint128 employeeId) external auth(REMOVE_EMPLOYEE_ROLE) {
-        /* check that employee exists */
-        require(employeeIds[employees[employeeId].accountAddress] != 0);
+    function terminateEmployee(
+        uint128 employeeId,
+        uint64 endDate
+    )
+        employeeExists(employeeId)
+        external
+        authP(TERMINATE_EMPLOYEE_ROLE, arr(uint256(employeeId)))
+    {
+        _terminateEmployee(employeeId, endDate);
+    }
 
-        // Pay owed salary to employee
-        // get time that has gone by (seconds)
-        uint256 time = getTimestamp().sub(employees[employeeId].lastPayroll);
-        if (time > 0)
-            _payTokens(employeeId, time);
+    /**
+     * @notice Adds `amount` to accrued value for employee with id `employeeId`
+     * @param employeeId Id of the employee
+     * @param amount Added amount
+     */
+    function addAccruedValue(
+        uint128 employeeId,
+        uint256 amount
+    )
+        employeeExists(employeeId)
+        external
+        authP(ADD_ACCRUED_VALUE_ROLE, arr(uint256(employeeId), amount))
+    {
+        require(amount <= MAX_ACCRUED_VALUE);
```

Comment:
I don't think so: if I recall correctly, the point of this check was to avoid the employer "blocking" the employee setting an accrued value so high that would make payroll overflow. But if an employee is owed more than `2^128`, it should be allowed to be paid.